nixos/testing: Switch from black to pyflakes

aszlig opened the pull request:
So far, we have used "black" for formatting the test code, which is rather strict and opinionated and when used inline in Nix expressions it creates all sorts of trouble.

One of the main annoyances is that when using strings coming from Nix expressions (eg. store paths or option definitions from NixOS modules), completely unrelated changes could cause tests to fail, since eg. black wants lines to be broken.

Another downside of enforcing a certain kind of formatting is that it makes the Nix expression code inconsistent because we're mixing two spaces of indentation (common in nixpkgs) with four spaces of indentation as defined in PEP-8. While this is perfectly fine for standalone Python files, it really looks ugly and inconsistent IMO when used within Nix strings.

What we actually want though is a linter that catches problems early on before actually running the test, because this is *actually* helping in development because running the actual VM test takes much longer.

This is the reason why I switched from black to pyflakes, because the latter actually has useful checks, eg. usage of undefined variables, invalid format arguments, duplicate arguments, shadowed loop vars and more.

Closes #72964

aszlig committed:
nixos/testing: Switch from black to pyflakes

So far, we have used "black" for formatting the test code, which is
rather strict and opinionated and when used inline in Nix expressions it
creates all sorts of trouble.

One of the main annoyances is that when using strings coming from Nix
expressions (eg. store paths or option definitions from NixOS modules),
completely unrelated changes could cause tests to fail, since eg. black
wants lines to be broken.

Another downside of enforcing a certain kind of formatting is that it
makes the Nix expression code inconsistent because we're mixing two
spaces of indentation (common in nixpkgs) with four spaces of
indentation as defined in PEP-8. While this is perfectly fine for
standalone Python files, it really looks ugly and inconsistent IMO when
used within Nix strings.

What we actually want though is a linter that catches problems early on
before actually running the test, because this is *actually* helping in
development because running the actual VM test takes much longer.

This is the reason why I switched from black to pyflakes, because the
latter actually has useful checks, eg. usage of undefined variables,
invalid format arguments, duplicate arguments, shadowed loop vars and
more.

Signed-off-by: aszlig <aszlig@nix.build>
Closes: https://github.com/NixOS/nixpkgs/issues/72964

roberth committed:
nixos/testing-python.nix: Expose driver

(cherry picked from commit a2c9220568648b4528154ebd8e657add243ed0b4)

roberth commented:
I was working on flake8 and I wrote some code to build the scripts only. See https://github.com/NixOS/nixpkgs/pull/122204

Do you think pyflakes is better than flake8? Or perhaps we should have both?

aszlig commented:
> Do you think pyflakes is better than flake8? Or perhaps we should have both?

From https://gitlab.com/pycqa/flake8/-/blob/master/README.rst#flake8:

> Flake8 is a wrapper around these tools:
> 
>     PyFlakes
>     pycodestyle
>     Ned Batchelder's McCabe script

The `pycodestyle` part is the part about enforcing PEP-8, so if we have pyflakes, we should have the most useful parts.

roberth committed:
nixos/testing: Set up scope for testScript linter

Our test driver exposes a bunch of variables and functions, which
pyflakes doesn't recognise by default because it assumes that the test
script is executed standalone. In reality however the test driver script
is using exec() on the testScript.

Fortunately pyflakes has $PYFLAKES_BUILTINS, which are the attributes
that are globally available on all modules to be checked. Since we only
have one module, using this environment variable is fine as opposed to
my first approach to this, which tried to use the unstable internal API
of pyflakes.

The attributes are gathered by the main derivation of the test driver,
because we don't want to end up defining a new attribute in the test
driver module just to being confused why using it in a test will result
in an error.

Another way we could have gathered these attributes would be in
mkDriver, which is where the linting takes place. However, we do have a
different set of Python dependencies in scope and duplicating these will
again just cause confusion over having it at one location only.

Signed-off-by: aszlig <aszlig@nix.build>
Co-Authored-By: aszlig <aszlig@nix.build>

roberth commented:
I've added some processing to define the variables that are defined externally. At least sufficient for the `nginx` and `sudo` tests.

roberth commented:
First batch, up to and including `a`

@GrahamcOfBorg test _3proxy acme agda airsonic amazon-init-shell ammonite atd avahi avahi-with-resolved awscli

roberth commented:
@GrahamcOfBorg test babeld bat bazarr bcachefs beanstalkd bees bind bitcoind bittorrent bitwarden blockbook-frontend boot boot-stage1 borgbackup buildbot buildkite-agents

aszlig commented on nixos/lib/testing-python.nix:
```diff
@@ -161,7 +161,12 @@ rec {
             ${lib.optionalString (!skipLint) ''
-              ${python3Packages.pyflakes}/bin/pyflakes $out/test-script
+              ( # Define the variables
+                echo "def testScript(subtest, start_all, ${lib.concatStringsSep ", " nodeNames}):"
```
Hm, I wonder whether there is a more elegant way to define this, since this could case all kinds of trouble we probably aren't aware of right now. One (pretty bad) example would be if you were doing something like this:

```
{
  testScript = ''
    machine.succeed('something
      where spaces are important')
  '';
}

roberth replied:
We need to do something. Defining them as variables might also work, but a linter may think too much of those variables.

We'll see. It doesn't alter the final script, so it can only cause a false error in the linter.

aszlig replied:
One issue with this is that line numbers are no longer accurate.

aszlig replied:
Okay, if we'd use `pyflakes.checker.Checker` directly, we should be able to directly provide a scope. Trying...

aszlig replied:
@roberth: After digging through the code, I've found `PYFLAKES_BUILTINS`, which IMO is a better solution. I've added this as a squash-commit.

roberth commented:
Ok, ofborg was a bad idea for this PR.
I'm testing this now:
```
NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1 nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "a" k && v?driver then v.driver else null) pkgs.nixosTests'
```

Still need to watch out for tests in nested attrs though.

andir commented on nixos/lib/testing-python.nix:
```diff
@@ -159,7 +161,12 @@ rec {
+                echo "def testScript(subtest, start_all, ${lib.concatStringsSep ", " nodeNames}):"
+                sed -e 's/^/    /' -e 's/^    $//' <$out/test-script
+              ) >test-script
+              ${python3Packages.pyflakes}/bin/pyflakes test-script
```
Can we move the actual linting into build that is detached from the rest? It is gaining a lot of complexity (comparatively).

Probably good to have a build for the script that passes if linting succeeded (unless disabled?).

roberth replied:
It's already possible to run only the linting

```
nix-build -A nixosTests.acme.driver
```

If that's all you want, it's done in this PR (or #122204 which is cherry-picked here).

If you want to have to attributes, one for running the test and one for linting, that's a new requirement that I would prefer to leave out of scope for now.

roberth committed:
nixosTests.acme: lint

roberth commented:
# Prefixes tested.

 - `gh pr checkout #122201`
 - pick a random letter
 - edit your github handle next to it (in this comment)
 - run nix-build
 - report or push any fixes

The command to run is as follows. Replace `XXXXX` by a letter.

```
NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1 nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "XXXXX" k && v?driver then v.driver else null) pkgs.nixosTests'
```

If you don't have privileges on the nixpkgs repo, leave a comment of your own instead.

 - [x] _ @roberth
 - [x] a @roberth
 - [ ] b
 - [ ] c
 - [ ] d
 - [ ] e
 - [ ] f
 - [ ] g
 - [ ] h
 - [ ] i
 - [ ] j
 - [ ] k
 - [ ] l
 - [ ] m
 - [ ] n
 - [ ] o
 - [ ] p
 - [ ] q
 - [ ] r
 - [ ] s
 - [ ] t
 - [ ] u
 - [ ] v
 - [ ] w
 - [ ] x
 - [ ] y
 - [ ] z

aszlig commented:
@roberth: Hydra jobset: https://headcounter.org/hydra/jobset/aszlig/nixos-tests

roberth commented on nixos/lib/testing-python.nix:
```diff
@@ -159,7 +171,10 @@ rec {
             ${lib.optionalString (!skipLint) ''
-              ${python3Packages.black}/bin/black --check --diff $out/test-script
+              PYFLAKES_BUILTINS="$(
+                echo -n ${lib.escapeShellArg (lib.concatStringsSep "," nodeNames)},
```
```suggestion
                echo -n ${lib.escapeShellArg (lib.concatStringsSep "," nodeHostNames)},
```

Subtle but important for amazon-init-shell and cloud-init.

aszlig replied:
Eeek, so this essentially means that you can create colliding nodes?

roberth replied:
It's not great. Having two namespace is nice though if you want hostnames with `.`'s in them, but that's not even supported now. Something for later.

blaggacao commented:
While running one build, I still see a spurious:

```console
running tests
Success: no issues found in 1 source file(B
All done! ✨ 🍰 ✨
1 file would be left unchanged.
```

It would be nice (to have), if you could make some well styled message to give a little better context to it, otherwise, succeeded without any error:

```
MYPREFIX=d NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1 nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
```

I've seen those, without any context. So it's unclear whether they are related and what this is supposed to tell me.
```console
building '/nix/store/pp9npy97mjmsz2mj219r06v47mdq8c7j-nginx.conf.drv'...

==================== Results ===================
No issues found.

==================== Summary ===================
Total issues:
    Unspecified: 0
    Low: 0
    Medium: 0
    High: 0

building '/nix/store/wjv3phkk03v2gv0pnkj4h1qsyd4qg5jb-nginx.conf.drv'...

==================== Results ===================
No issues found.

==================== Summary ===================
Total issues:
    Unspecified: 0
    Low: 0
    Medium: 0
    High: 0

```

blaggacao commented:
```console
MYPREFIX=e NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
[1]    96595 segmentation fault (core dumped)  nix-build --expr
```

roberth commented:
@blaggacao Are you running `nixUnstable`? That one is known to be unstable in that sense as well. `nix-shell -p nix` probably won't help because of a change in the protocol that breaks logging to Nix 2.3...

`nix-build` isn't _great_ for concurrent logs, so you may want to take a clue from the final line (which drvs failed) and work from there.

blaggacao commented:
```console
$ MYPREFIX=f NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
/nix/store/94437yb523wb2sj4g49v6r2cjvya1dbz-nixos-test-driver-fancontrol
/nix/store/9wfrvjhfix0fzjmp2zw2lsw0wr044dzi-nixos-test-driver-fcitx
/nix/store/idfhyz8rkdyivnqprbxjycxhablnrmva-nixos-test-driver-ferm
/nix/store/msr6nlhgx0cvb55sbl9rwyb930gm2095-nixos-test-driver-firefox
/nix/store/x1g3i1vnjfgn34cggbs23sg2c7lwhxpf-nixos-test-driver-firefox
/nix/store/zn8dqqv4gi30qz7lsfc6xnrsqzb45sna-nixos-test-driver-firejail
/nix/store/kcgjirrvv8j423iqhnnv1sii87jg1n5n-nixos-test-driver-firewall
/nix/store/3frn2gddlxxmmvvaddgghd0l6kq1y0k0-nixos-test-driver-fish
/nix/store/3j0lk5cczl7g79i60zw6kr2p339bn7c8-nixos-test-driver-flannel
/nix/store/xnylfrgb1nvpphnb8am899zgy2i6ks1s-nixos-test-driver-fluentd
/nix/store/khckl96vgjkdnkf3v0vjxknc0xxm8aw8-nixos-test-driver-fontconfig-default-fonts
/nix/store/cv5iqi40l02g3cy3yjsgv26nlaz5pycl-nixos-test-driver-freeswitch
/nix/store/kswkxqlpsh32my33ymrzq1w9y5q88xs4-nixos-test-driver-fsck
/nix/store/k2ibva8k8fd88yw47l3wlr7gq71bjjim-nixos-test-driver-ft2-clone
```

(no issues with "f")

blaggacao commented:
```console
MYPREFIX=g NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
[ ... ]
created 1250 symlinks in user environment
nix-build: src/libstore/build/local-derivation-goal.cc:2460: virtual void nix::LocalDerivationGoal::registerOutputs(): Assertion `newInfo.ca' failed.
[1]    133728 abort (core dumped)  nix-build --expr
```

knedlsepp commented:
> ```
> MYPREFIX=g NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
> [ ... ]
> created 1250 symlinks in user environment
> nix-build: src/libstore/build/local-derivation-goal.cc:2460: virtual void nix::LocalDerivationGoal::registerOutputs(): Assertion `newInfo.ca' failed.
> [1]    133728 abort (core dumped)  nix-build --expr
> ```

This seems unrelated and was reported here
https://github.com/NixOS/nix/issues/4256

blaggacao commented:
```console
$ MYPREFIX=i NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
[ ... ]
/nix/store/4n8m2q9y2rkzv7v0dlkifqzmz7dnfv07-nixos-test-driver-i3wm
/nix/store/iczgjsiwf01lqpy1hkrcpg39b9ah0br5-nixos-test-driver-icingaweb2
/nix/store/ykdwyg7cvzib5ynkx7lf4iysiwgjlfx8-nixos-test-driver-iftop
/nix/store/h2mcbs9piw1imxzsv3hw607avxjmfb74-nixos-test-driver-ec2-image-contents
/nix/store/nckh64bdy8wf5p690v2d759zhbla071j-nixos-test-driver-incron
/nix/store/j1f86v77228kc08a0gargaqrf1vcm10i-nixos-test-driver-influxdb
/nix/store/qzr1ls0zlicqdpx437r90nff96xpqfsw-nixos-test-driver-initrd-network-openvpn
/nix/store/4zv8d3npygd1vvqnppqw5c5c27sw361j-nixos-test-driver-initrd-network-ssh
/nix/store/y0242a8hbjlgblwpyzwn4f6rklian78s-nixos-test-driver-initrd-network
/nix/store/ymhn6ya2p9zyv62n223238061s6w587n-nixos-test-driver-inspircd
/nix/store/ydrks519pg241m0ivx3a81jlyb9ka0xs-nixos-test-driver-iodine
/nix/store/q8nk64wawpskap5sa1yq6jn5c6nbw4g9-nixos-test-driver-ipfs
/nix/store/br7fk605gz1ja4rz1yakd9r7advjprp5-nixos-test-driver-ipv6
/nix/store/7x3jx60586vqajb6a2kwjm12r1jkdkrr-nixos-test-driver-iscsi
```

Commit:
nixos/testing: lint jellyfin test

blaggacao commented:
With: https://github.com/xoe-labs/nixpkgs/commit/e5dd8fe94f3eadb0b0f63fb0afa386cb4dfe3bf2

```console
$ MYPREFIX=j NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
[ ... ]
/nix/store/2amzi8y2j2crf00ai0slyd9zq3pqsq63-nixos-test-driver-jackett
/nix/store/zqd9ahb7c92anl6dxb1v5xm0knpna3zn-nixos-test-driver-jellyfin
/nix/store/4icp3i2sk0lnmp8hwycdc6ahr3xyhfk0-nixos-test-driver-jenkins
/nix/store/j2wk9hjwsdymp68f8ja3nb34ahsyr1sq-nixos-test-driver-jirafeau
/nix/store/fk1piknvj6iv8zwb2s3237jb2j8ykh6m-nixos-test-driver-jitsi-meet
```

roberth committed:
nixos/tests/nfs: lint

Ma27 committed:
nixosTests.chromium: lint

Note: I didn't execute it entirely because I'd have to build chromium
for this, but the diff appears fine.

Ma27 committed:
nixosTests.containers-custom-pkgs: lint

The new linter basically does

   def testScript
      # ...

before calling `pyflakes`. As this test-script is empty, it would lead
to a syntax-error unless `pass` is added.

blaggacao commented:
```console
$ MYPREFIX=k NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
[ ... ]
/nix/store/8whvh81byhg7war8dsv2gb0ap3s9n0dl-nixos-test-driver-k3s
/nix/store/37wk3jx9xf7shrksg0mi4z4q0lxj2vhf-nixos-test-driver-keepalived
/nix/store/6071jcdf6p41njif4nn97763ijgnk933-nixos-test-driver-keepassxc
/nix/store/i2kcb4wj8k6yjxxmfm8219iz4whzwl48-nixos-test-driver-kernel-latest-ath-user-regd
/nix/store/9ljim9ywnf3n5vkgvnwbiqv5v2iqs89w-nixos-test-driver-knot
/nix/store/azwypfwdhxnmww87bdmmk8m1m0znk89g-nixos-test-driver-ksm
```

Ma27 committed:
nixosTests.containers-imperative: lint

Ma27 committed:
nixosTests.custom-ca: lint

blaggacao commented:
```console
$ MYPREFIX=l NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
[ ... ]
/nix/store/qjrrny1ba20w36jqak5jmra5agrl6rp5-nixos-test-driver-leaps
/nix/store/2jwlxf75iqd1zppj12ikip90yhzsjxsz-nixos-test-driver-lidarr
/nix/store/zldkhk28n8b4mww0df7y3z86vgsw7zkj-nixos-test-driver-lightdm
/nix/store/yksfli2n9awvyny0crlxzykz9jxgiffq-nixos-test-driver-limesurvey
/nix/store/c18m4p3q8y9d3k5a6pj59w7qd7bbrfsq-nixos-test-driver-locate
/nix/store/ac7lk9m2wdjxb39m1bxf2kv760z2lf3z-nixos-test-driver-login
/nix/store/p8zwiqi19465p2jq40kwibm8nw85ciip-nixos-test-driver-loki
/nix/store/0q5rmb28pbgqaka9d5i8xqqmapiw7ak1-nixos-test-driver-unnamed
/nix/store/ba0nz94pa2qvlxli1xc4h5ll5yvbrma0-nixos-test-driver-lsd
/nix/store/xl57pyj98x7jir6xa482w3grmac5lils-nixos-test-driver-lxd
/nix/store/4fnyracah9czj61v89xjfaja5svg9fpa-nixos-test-driver-lxd-nftables
```

Ma27 commented:
IIRC yes because their own license is unfree.

aszlig committed:
nixos/tests/yggdrasil: Fix linting error

Linter error was: f-string is missing placeholders

Signed-off-by: aszlig <aszlig@nix.build>

blaggacao commented:
Interim, please pull in:
- https://github.com/xoe-labs/nixpkgs/commit/20274bb562a0845a92eb2f1c6a050760a6d04147
- https://github.com/xoe-labs/nixpkgs/commit/e5dd8fe94f3eadb0b0f63fb0afa386cb4dfe3bf2

aszlig commented:
> Interim, please pull in:
> 
>     * [xoe-labs@20274bb](https://github.com/xoe-labs/nixpkgs/commit/20274bb562a0845a92eb2f1c6a050760a6d04147)
> 
>     * [xoe-labs@e5dd8fe](https://github.com/xoe-labs/nixpkgs/commit/e5dd8fe94f3eadb0b0f63fb0afa386cb4dfe3bf2)

Why are these commits marked with fixup?

blaggacao commented:
> Why are these commits marked with fixup!?

User error. :innocent:

aszlig committed:
nixos/tests/networking: Fix str literal comparison

Linter error:

  use ==/!= to compare constant literals (str, bytes, int, float, tuple)

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
@blaggacao: Please fix up the `fixup` commits and I'll cherry-pick them in :-)

aszlig committed:
nixos/tests/printing: Remove unused 'sys' import

Signed-off-by: aszlig <aszlig@nix.build>

blaggacao commented:
```console
$ MYPREFIX=n NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests'
[ ... ]
/nix/store/f5l1r8zp4xngsl8frgc4zq49gs3616r3-nixos-test-driver-n8n
/nix/store/jkgjlkpf47q79a8djbs3sfm8405mv4g5-nixos-test-driver-nagios
/nix/store/b5xisw99x4nqs159xc6wflbflxhiwyyr-nixos-test-driver-nano
/nix/store/50sz41vi14f0426hyyfz4ql71ii0d3hl-nixos-test-driver-nar-serve
/nix/store/m2r42fyr1llhszwq5aay014033id7zqj-nixos-test-driver-ncdns
/nix/store/528qrw9x4kngfkv7khzg2hrv48zqghhb-nixos-test-driver-ndppd
/nix/store/328v66xshdmv3910k3p5l4kkpac1cxlz-nixos-test-driver-nebula
/nix/store/yhm8lfj0qq7l02cmfn5rdzm9mg0crpmz-nixos-test-driver-neo4j
/nix/store/lkkylsc4chlx6kaf829r057ir61s9kgc-nixos-test-driver-netdata
/nix/store/70gmbi5f0f5nshl9yzwbyyas4fkc36vb-nixos-test-driver-networking-proxy
/nix/store/6qlgkv6qyld7bfkpa0kf179spq2ds4yy-nixos-test-driver-nexus
/nix/store/3pd89czwipl1nravln26xs4n3sb7awxv-nixos-test-driver-nghttpx
/nix/store/ciapyz7s7y2y4b5b0wyi0y4izpziwnrs-nixos-test-driver-nginx
/nix/store/vxyxwzcbadxi6i1g9c48m3wllr142yn5-nixos-test-driver-nginx-auth
/nix/store/9psvypnbldj1924l39raj9ca4368n3k8-nixos-test-driver-nginx-etag
/nix/store/2qb2ixky2r4hf2xjpx9lbpaqdca10678-nixos-test-driver-nginx-pubhtml
/nix/store/x86rmqcrw7nlqlgbc4bh8jygxpblz27b-nixos-test-driver-nginx-sandbox
/nix/store/fxrhwxa0c45chms4sxgaarqbkwlilvj6-nixos-test-driver-nginx-sso
/nix/store/k1wpg22ykz60vq5avvdk205y9d82mnph-nixos-test-driver-nix-ssh-serve
/nix/store/f2zpn6ipif513z2l7bmdmhvmj2vrq5nq-nixos-test-driver-nixos-generate-config
/nix/store/qgnp6vs557fl1zlsl0hai7zfs9hdjipb-nixos-test-driver-nomad
/nix/store/ycdb9f5wk0s6r843sqrbw6i1rh8qz44q-nixos-test-driver-novacomd
/nix/store/27i7c4skzh5q34qx0x6dmacmxzf6b46r-nixos-test-driver-nsd
/nix/store/p03k4x7b3llfp64k7zfqa9b9hp29h4p8-nixos-test-driver-nzbget
/nix/store/i0vxkv7s41imq9ikdhv3jsxdszqji6yn-nixos-test-driver-nzbhydra2
```

blaggacao commented:
- https://github.com/xoe-labs/nixpkgs/commit/88abb3605bad0fa2146aaa95ccdc736d85bf992a
- https://github.com/xoe-labs/nixpkgs/commit/b33e0d4304dc05635783e0c87e8e0f3729838c60

aszlig committed:
nixos/tests/shadow: Fix linting errors

Linter errors reported:

  6:32 f-string is missing placeholders
  7:26 f-string is missing placeholders
  8:32 f-string is missing placeholders
  30:32 f-string is missing placeholders
  31:26 f-string is missing placeholders
  32:32 f-string is missing placeholders
  48:32 f-string is missing placeholders
  49:26 f-string is missing placeholders
  50:32 f-string is missing placeholders
  76:32 f-string is missing placeholders
  77:26 f-string is missing placeholders
  78:32 f-string is missing placeholders

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
>   * [xoe-labs@88abb36](https://github.com/xoe-labs/nixpkgs/commit/88abb3605bad0fa2146aaa95ccdc736d85bf992a)

This commit seems unrelated to this PR, can you please open a new PR for this please?

aszlig committed:
nixos/tests/unbound: Remove unused 'json' import

Signed-off-by: aszlig <aszlig@nix.build>

blaggacao commented:
For "u", there is: https://github.com/NixOS/nixpkgs/issues/122257

blaggacao commented:
```console
$ MYPREFIX=v NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests' --show-trace
[ ... ]
/nix/store/cdpwck9m5ggdyv4yg0r3c90wk9ywy76y-nixos-test-driver-v2ray
/nix/store/sc39sx0b79fc8w099w407qb6lz33rb7c-nixos-test-driver-vault
/nix/store/xa71irdpz7bwlg7768hbvh77kl557mjp-nixos-test-driver-vault-postgresql
/nix/store/0b1bnshdl8r2dayqd0fbbxj4q54scsm8-nixos-test-driver-victoriametrics
/nix/store/mnfnp9d5629r5ck2l7g0953855mdflfr-nixos-test-driver-vscodium
```

aszlig committed:
nixos/test/virtualbox: Fix linting errors

There were a bunch of unnecessary f-strings in there and I also removed
the "# fmt: on/off" comments, because we no longer use Black and thus
won't need those comments anymore.

Signed-off-by: aszlig <aszlig@nix.build>

blaggacao commented:
- except for: https://github.com/NixOS/nixpkgs/issues/122257
- with: https://github.com/xoe-labs/nixpkgs/commit/26b0c2a5578c51a0e014daf79febb8356e5beb11 (please cherry-pick)

```console
$ MYPREFIX=u NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests' --show-trace
[ ... ]
/nix/store/1r16rvb06nmzfqsfv6giyqgq1i9x4jw2-nixos-test-driver-ucg
/nix/store/v2714kvxqlwq52hsqhza89qw358xl6qb-nixos-test-driver-udisks2
/nix/store/fq63axq30skvx3ffd3s43vf73mwszgci-nixos-test-driver-unbound
/nix/store/7gkidkv91viyddlx2z70snqfm2hkcfkh-nixos-test-driver-upnp
/nix/store/8m5n1sgqq7pgc4vh9lrxjp1i8kvalf3r-nixos-test-driver-usbguard
/nix/store/qrf89q55jbkmkj412gy6b3adrmbb03h8-nixos-test-driver-uwsgi
```

aszlig commented:
>     * with: [xoe-labs@26b0c2a](https://github.com/xoe-labs/nixpkgs/commit/26b0c2a5578c51a0e014daf79febb8356e5beb11) (please cherry-pick)

Already fixed in b296553f40ba8939f9f143a9006da21d186982b2, please update your branch.

blaggacao commented:
```console
$ MYPREFIX=q NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests' --show-trace
[ ... ]
/nix/store/3r4xy6ls10g6aikc6zgcs6hjyr9sx9hi-nixos-test-driver-qboot
/nix/store/jasmbny696k9ibgh4hk88mhpjlrj5wbp-nixos-test-driver-quorum
```

blaggacao commented:
```
swtiching off:
  # hocker-fetchdocker = handleTest ./hocker-fetchdocker {};
  # home-assistant = handleTest ./home-assistant.nix {};
```
"h" succeeds

```console
$ MYPREFIX=h NIXPKGS_ALLOW_INSECURE=1 NIXPKGS_ALLOW_UNFREE=1; nix-build --expr 'let pkgs = import ./. { config.allowUnfree = true; }; in pkgs.lib.mapAttrs (k: v: if pkgs.lib.strings.hasPrefix "'${MYPREFIX}'" k && v?driver then v.driver else null) pkgs.nixosTests' --show-trace
[ ... ]
/nix/store/6ppglyakg4xar9mgcn0ihl7ws3wv163x-nixos-test-driver-handbrake
/nix/store/dgr7x46qvplq11gmbx22ajksidkx7jrx-nixos-test-driver-haproxy
/nix/store/60zp0bkw83bs2rcc8vhvcr4rsk5kpzal-nixos-test-driver-hardened
/nix/store/w187li3pm4xhf3bbdpzy5556x93xd5sp-nixos-test-driver-hedgedoc
/nix/store/adkyr1prxl5zg4y003nzagfx51wzmp9a-nixos-test-driver-herbstluftwm
/nix/store/b4dadhsxnr28d2vxq2wgmwp5yi9b8dxv-nixos-test-driver-hibernate
/nix/store/zq2g6rqsykl90s97fwcq0q4asllrcfsb-nixos-test-driver-hitch
/nix/store/7zpsk1hify137lsz5dbng5mcs2fmzcrb-nixos-test-driver-hledger-web
/nix/store/rql6xp5awmbp4h7i77p33nwfqqzir93v-nixos-test-driver-hound
/nix/store/gw5zhdnb4c8z2zwgqa5w91lg5zlb5v43-nixos-test-driver-hub
```

blaggacao commented:
OK, I'm out. Thanks for this initiative! :heart:

aszlig commented:
> OK, I'm out. Thanks for this initiative! heart

Thanks a lot for helping with this, good night :-)

aszlig commented:
FYI: I've changed the Hydra jobset to only evaluated the `.driver` attribute, so we only get linting errors and other build errors leading up to the test but without running the tests themselves.

roberth reviewed (approved):
~Don't merge yet, while aszlig is doing some final touches.~



#### General remarks

- We improved over a dozen of test scripts
- Most tests already passed linting, so this linter can hardly be a burden

#### Testing

With our manual testing, we should have covered all test scripts. Some test scripts were not testable because of failed dependencies or evaluation errors. This is ok, because the goal is to prevent _new_ bitrot and avoid surprises and small chores where possible.

#### Conclusion

This PR achieves the goal of https://github.com/NixOS/nixpkgs/issues/72964 with minimal disruption.
